### DISCUSSION TITLE
Is the coding error when changing milvus1.1.2 to milvus2.4?

### DISCUSSION THREAD

**Discussion by watertianyi** [milvus-io/milvus#31748] [Q&A and General discussion]
The source code is in
https://github.com/PaddlePaddle/PaddleNLP/blob/develop/applications/text_classification/multi_class/retrieval_based/utils/milvus_util.py

I modified：https://github.com/PaddlePaddle/PaddleNLP/issues/8215
Can you help me take a look? What went wrong?

**Comment by yhmo** [comment]
```
# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from config import (
    MILVUS_HOST,
    MILVUS_PORT,
    collection_param,
    index_param,
    index_type,
    search_param,
    top_k,
)

from pymilvus import (
     connections,
     utility,
     FieldSchema, CollectionSchema, DataType,
     Collection,
 )


class VecToMilvus:
    def __init__(self):
        connections.connect(host=MILVUS_HOST, port=MILVUS_PORT)

    def has_collection(self, collection_name):
        try:
            ok = utility.has_collection(collection_name)
            return ok
        except Exception as e:
            print("Milvus has_table error:", e)

    def creat_collection(self, collection_name):
        try:
            field_1 = FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False)
            field_2 = FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=dim)
            schema = CollectionSchema(fields=[field_1, field_2])

            collection = Collection(collection_name, schema)
            print("collection created")

            return True
        except Exception as e:
            print("Milvus create collection error:", e)

    def create_index(self, collection_name):
        try:
            collection = Collection(collection_name)
            vector_index = {
                'metric_type': "L2",
                'index_type': "IVF_FLAT",
                'params': {"nlist": 1024}
            }
            collection.create_index("vector", vector_index)
            collection.load()
            return True
        except Exception as e:
            print("Milvus create index error:", e)

    def has_partition(self, collection_name, partition_tag):
        try:
            ok = utility.has_collection(collection_name)
            if not ok:
                return ok

            ok = utility.has_partition(collection_name=collection_name, partition_name=partition_tag)
            return ok
        except Exception as e:
            print("Milvus has partition error: ", e)

    def delete_partition(self, collection_name, partition_tag):
        try:
            collection = Collection(collection_name)
            collection.drop_partition(partition_name=partition_tag)
            return True
        except Exception as e:
            print("Milvus has partition error: ", e)

    def create_partition(self, collection_name, partition_tag):
        try:
            collection = Collection(collection_name)
            collection.create_partition(partition_name=partition_tag)
            print("create partition {} successfully".format(partition_tag))
            return True
        except Exception as e:
            print("Milvus create partition error: ", e)

    def insert(self, vectors, collection_name, ids=None, partition_tag=None):
        try:
            if not self.has_collection(collection_name):
                self.creat_collection(collection_name)
                self.create_index(collection_name)
                print("collection info: {}".format(self.client.get_collection_info(collection_name)[1]))
            if (partition_tag is not None) and (not self.has_partition(collection_name, partition_tag)):
                self.create_partition(collection_name, partition_tag)

            collection = Collection(collection_name)
            data = [
                ids,
                vectors,
            ]
            collection.insert(data)
            print(
                "Insert {} entities, there are {} entities after insert data.".format(
                    len(ids), self.client.count_entities(collection_name)[1]
                )
            )
            return status, ids
        except Exception as e:
            print("Milvus insert error:", e)


class RecallByMilvus:
    def __init__(self):
        self.client = Milvus(host=MILVUS_HOST, port=MILVUS_PORT)

    def search(self, vectors, collection_name, partition_tag=None):
        try:
            collection = Collection(collection_name)
            partition_names = [] if partition_tag is None else [partition_tag]
            results = collection.search(
                data=vectors,
                anns_field="vector",
                partition_names=partition_names,
                limit=top_k,
            )

            return True, results
        except Exception as e:
            print("Milvus recall error: ", e)
```

**Reply by yhmo** [reply]
先把milvus 2.x的api搞会再来换，不要什么都不看就换
https://milvus.io/docs/manage-collections.md

**Reply by watertianyi** [reply]
Milvus recall error:  <MilvusException: (code=101, message=failed to search: collection not loaded

**Reply by yhmo** [reply]
Call the methods in this sequence:  creat_collection() -----> create_index() ------> insert() ---> insert() ---> ... ---> search()

**Comment by yhmo** [comment]
请各位提问时尽量使用英语，Milvus社区里有世界各地的工程师，用英文问答所有人都能看得懂。

**Comment by watertianyi** [comment]
The code is modified as follows. It runs correctly, but the results are incorrect and the retrieval accuracy is reduced.
from config import (
    MILVUS_HOST,
    MILVUS_PORT,
    top_k,
)

from pymilvus import (
     connections,
     utility,
     FieldSchema, CollectionSchema, DataType,
     Collection,Partition
 )


class VecToMilvus:
    def __init__(self):
        connections.connect(host='localhost', port='19530')
        # self.client2 = MilvusClient("http://localhost:19530")
        print(f"\nList connections:")
        print(connections.list_connections())
    def has_collection(self, collection_name):
        try:
            ok = utility.has_collection(collection_name)
            return ok
        except Exception as e:
            print("Milvus has_table error:", e)

    def creat_collection(self, collection_name):
        try:
            field_1 = FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False)
            field_2 = FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=768)
            schema = CollectionSchema(fields=[field_1, field_2])

            collection = Collection(collection_name, schema)
            # collection.load()
            print("collection created")

            return True
        except Exception as e:
            print("Milvus create collection error:", e)

    def create_index(self, collection_name):
        try:
            collection = Collection(collection_name)
            vector_index = {
                'metric_type': "L2",
                'index_type': "IVF_FLAT",
                'params': {"nlist": 1000}
            }
            collection.create_index("vector", vector_index)
            collection.load()
            return True
        except Exception as e:
            print("Milvus create index error:", e)

    def has_partition(self, collection_name, partition_tag):
        try:
            ok = utility.has_collection(collection_name)
            if not ok:
                return ok

            ok = utility.has_partition(collection_name=collection_name, partition_name=partition_tag)
            return ok
        except Exception as e:
            print("Milvus has partition error: ", e)

    def delete_partition(self, collection_name, partition_tag):
        try:
            print("delete.......")
            collection = Collection(collection_name)
            partition = Partition(collection, name=partition_tag)     # Get an existing partition.
            partition.release()
            partition.drop()
            collection.drop_partition(partition_name=partition_tag)
            collection.release()
            collection.drop()

            return True
        except Exception as e:
            print("Milvus has partition error: ", e)

    def create_partition(self, collection_name, partition_tag):
        try:
            
            collection = Collection(collection_name)
            collection.create_partition(partition_name=partition_tag)
            print("create partition {} successfully".format(partition_tag))
            return True
        except Exception as e:
            print("Milvus create partition error: ", e)

    def insert(self, vectors, collection_name, ids=None, partition_tag=None):
        try:
            if not self.has_collection(collection_name):
                self.creat_collection(collection_name)
                self.create_index(collection_name)
                print("collection info: {}".format(self.client.get_collection_info(collection_name)[1]))
            if (partition_tag is not None) and (not self.has_partition(collection_name, partition_tag)):
                self.create_partition(collection_name, partition_tag)

            collection = Collection(collection_name)
            data = [
                ids,
                vectors,
            ]
            partition = Partition(collection, name=partition_tag)
            partition.insert(data)
            # collection.insert(data)
            print("type:",len(data[1]))
            collection.flush()
            print(
                "Insert {} entities, there are {} entities after insert data.".format(
                    len(data[1]), collection.num_entities
                )
            )
            return data
        except Exception as e:
            print("Milvus insert error:", e)


class RecallByMilvus:
    def __init__(self):
        connections.connect(host=MILVUS_HOST, port=MILVUS_PORT)
        print(f"\nList connections:")
        print(connections.list_connections())

    def search(self, vectors, collection_name, partition_tag=None):
        try:
            collection = Collection(collection_name)
            
            partition_names = [] if partition_tag is None else [partition_tag]
            partition = Partition(collection, name=partition_tag)
            partition.load()
            search_params = {"metric_type": "L2", "params": {"nprobe": 20}}
            results = partition.search(vectors,"vector",search_params,top_k)
            # results = collection.search(
            #     data=vectors,
            #     anns_field="vector",
            #     param={"metric_type": "L2", "params": {"nprobe":20}},
            #     limit=top_k,
            # )
            print("results:",results)
            return results
        except Exception as e:
            print("Milvus recall error: ", e)

**Reply by yhmo** [reply]
Change this line: 
`search_params = {"metric_type": "L2", "params": {"nprobe": 20}}`

to be:
`search_params = {"metric_type": "L2", "params": {"nprobe": 1000}}`

then the result will be accurate.

**Reply by watertianyi** [reply]
The similarity is also very low

**Reply by yhmo** [reply]
When nprobe==nlist, the search is a brute-force search, the recall rate is 100%. If the result is below expectation, I think you can change your embedding model.
